src/sagemaker/tuner.py

```diff
@@ -0,0 +1,398 @@
+        super(IntegerParameter, self).__init__(min_value, max_value)
+
+
+class HyperparameterTuner(object):
+    TUNING_JOB_NAME_MAX_LENGTH = 32
+
+    SAGEMAKER_ESTIMATOR_MODULE = 'sagemaker_estimator_module'
+    SAGEMAKER_ESTIMATOR_CLASS_NAME = 'sagemaker_estimator_class_name'
+
+    DEFAULT_ESTIMATOR_MODULE = 'sagemaker.estimator'
+    DEFAULT_ESTIMATOR_CLS_NAME = 'Estimator'
+
+    def __init__(self, estimator, objective_metric_name, hyperparameter_ranges, metric_definitions=None,
+                 strategy='Bayesian', objective_type='Maximize', max_jobs=1, max_parallel_jobs=1,
+                 base_tuning_job_name=None):
+        self._hyperparameter_ranges = hyperparameter_ranges
+        if self._hyperparameter_ranges is None or len(self._hyperparameter_ranges) == 0:
+            raise ValueError('Need to specify hyperparameter ranges')
+
+        self.estimator = estimator
+        self.objective_metric_name = objective_metric_name
+        self.metric_definitions = metric_definitions
+        self._validate_parameter_ranges()
+
+        self.strategy = strategy
+        self.objective_type = objective_type
+        self.max_jobs = max_jobs
+        self.max_parallel_jobs = max_parallel_jobs
+
+        self.base_tuning_job_name = base_tuning_job_name
+        self._current_job_name = None
+        self.latest_tuning_job = None
+
+    def _prepare_for_training(self, job_name=None):
+        if job_name is not None:
+            self._current_job_name = job_name
+        else:
+            base_name = self.base_tuning_job_name or base_name_from_image(self.estimator.train_image())
+            self._current_job_name = name_from_base(base_name, max_length=self.TUNING_JOB_NAME_MAX_LENGTH, short=True)
+
+        self.static_hyperparameters = {to_str(k): to_str(v) for (k, v) in self.estimator.hyperparameters().items()}
+        for hyperparameter_name in self._hyperparameter_ranges.keys():
+            self.static_hyperparameters.pop(hyperparameter_name, None)
+
+        # For attach() to know what estimator to use for non-1P algorithms
+        # (1P algorithms don't accept extra hyperparameters)
+        if not isinstance(self.estimator, AmazonAlgorithmEstimatorBase):
+            self.static_hyperparameters[self.SAGEMAKER_ESTIMATOR_CLASS_NAME] = json.dumps(
+                self.estimator.__class__.__name__)
+            self.static_hyperparameters[self.SAGEMAKER_ESTIMATOR_MODULE] = json.dumps(self.estimator.__module__)
+
+    def fit(self, inputs, job_name=None, **kwargs):
+        """Start a hyperparameter tuning job.
+
+        Args:
+            inputs (str): Parameters used when called :meth:`~sagemaker.estimator.EstimatorBase.fit`.
+            job_name (str): Tuning job name. If not specified, the tuner generates a default job name,
+                based on the training image name and current timestamp.
+            **kwargs: Other arguments
+        """
+        if isinstance(inputs, list) or isinstance(inputs, RecordSet):
```

Comment:
This is much better:

```python
kwargs = dict(kwargs)
kwargs['job_name'] = job_name
self._prepare_for_training(**kwargs)
```

Reply:
basically replace line 140 to 145 with that block.

Reply:
`_prepare_for_training()` still needs `records` for 1P estimators but not for the others, though. Instead, it'd end up looking like:

```python
        kwargs = dict(kwargs)
        kwargs['job_name'] = job_name
        if isinstance(inputs, list) or isinstance(inputs, RecordSet):
            kwargs['records'] = inputs
        self.estimator._prepare_for_training(**kwargs)
```